jax/_src/profiler.py

```diff
@@ -70,8 +78,66 @@ def start_trace(log_dir):
       raise RuntimeError("Profile has already been started. "
                          "Only one profile may be run at a time.")
     _profile_state.profile_session = xla_client.profiler.ProfilerSession()
+    _profile_state.create_perfetto_link = create_perfetto_link
     _profile_state.log_dir = log_dir
 
+def _write_perfetto_trace_file(log_dir):
+  # Navigate to folder with the latest trace dump to find `trace.json.jz`
+  curr_path = os.path.abspath(log_dir)
+  root_trace_folder = os.path.join(curr_path, "plugins", "profile")
+  trace_folders = [os.path.join(root_trace_folder, trace_folder) for
+      trace_folder in os.listdir(root_trace_folder)]
+  latest_folder = max(trace_folders, key=os.path.getmtime)
+  trace_jsons = glob.glob(os.path.join(latest_folder, "*.trace.json.gz"))
+  if len(trace_jsons) != 1:
+    raise ValueError(f"Invalid trace folder: {latest_folder}")
+  trace_json, = trace_jsons
+
+  logging.info("Loading trace.json.gz and removing its metadata...")
+  # Perfetto doesn't like the `metadata` field in `trace.json` so we remove
+  # it.
+  # TODO(sharadmv): speed this up by updating the generated `trace.json`
+  # to not include metadata if possible.
+  with gzip.open(trace_json, "rb") as fp:
+    trace = json.load(fp)
+    del trace["metadata"]
+  filename = "perfetto_trace.json.gz"
+  perfetto_trace = os.path.join(latest_folder, filename)
+  logging.info("Writing perfetto_trace.json.gz...")
+  with gzip.open(perfetto_trace, "w") as fp:
+    fp.write(json.dumps(trace).encode("utf-8"))
+  return perfetto_trace
+
+class _PerfettoServer(http.server.SimpleHTTPRequestHandler):
+  """Handles requests from `ui.perfetto.dev` for the `trace.json`"""
+
+  def end_headers(self):
+    self.send_header('Access-Control-Allow-Origin', '*')
+    return super().end_headers()
+
+  def do_GET(self):
+    self.server.last_request = self.path
+    return super().do_GET()
+
+  def do_POST(self):
+    self.send_error(404, "File not found")
+
+def _host_perfetto_trace_file(log_dir):
+  # ui.perfetto.dev looks for files hosted on `127.0.0.1:9001`. We set up a
+  # TCP server that is hosting the `perfetto_trace.json.gz` file.
+  port = 9001
+  abs_filename = _write_perfetto_trace_file(log_dir)
+  directory, filename = os.path.split(abs_filename)
+  os.chdir(directory)
+  socketserver.TCPServer.allow_reuse_address = True
+  with socketserver.TCPServer(('127.0.0.1', port), _PerfettoServer) as httpd:
+    url = f"https://ui.perfetto.dev/#!/?url=http://127.0.0.1:{port}/{filename}'"
+    print(f"Open URL in browser: {url}")
+
+    # Once ui.perfetto.dev acquires trace.json from this server we can close
+    # it down.
+    while httpd.__dict__.get('last_request') != '/' + filename:
```

Comment:
It's kind of unfortunate that this is blocking until the user clicks the URL, and then the URL is invalid after that. I'm not sure if there's a better way to do it though. Can you mention this in the documentation, and also maybe mention that the TB-based profiling can save traces without user interaction?